File: banda/SergioDeIscar-banda-java/SergioDeIscar-banda-java/src/main/java/org/example/factories/MusicianFactory.java
package org.example.factories;

import org.example.enums.GuitarraType;
import org.example.enums.PercussionType;
import org.example.enums.TonoType;
import org.example.models.*;

import java.util.Arrays;

public class MusicianFactory {
    private static MusicianFactory persona;

    public static MusicianFactory getInstance() {
        if (persona == null) {
            persona = new MusicianFactory();
        }
        return persona;
    }

    private String getNombreRandom(){
        String[] names = {"Pepe", "Juan", "Ana", "Pedro", "María"};
        return names[(int)(Math.random() * names.length)];
    }

    private TonoType getTonoRandom(){
        return TonoType.values()[(int)(Math.random() * TonoType.values().length)];
    }

    private GuitarraType getGuitarraRandom(){
        return GuitarraType.values()[(int)(Math.random() * GuitarraType.values().length)];
    }

    private PercussionType getPercussionRandom(){
        return PercussionType.values()[(int)(Math.random() * PercussionType.values().length)];
    }


    public Musician[] getBandaRandom(int[] porcentajes){
        if (porcentajes == null){
            porcentajes = new int[] {20, 10, 10, 10, 15, 15, 15, 5};
        }
        if (porcentajes.length != 8) throw new IllegalArgumentException("Error: porcentajes ha de tener un tamaño de 8");
        for (var item : porcentajes){
            if (item == 0)
                throw new IllegalArgumentException("Error: ninguno valor de porcentajes puede sr menor o igual a 0");
        }
        if (Arrays.stream(porcentajes).sum() != 100)
            throw new IllegalArgumentException("Error: la suma de todos los valores de porcentajes a de ser igual a 100");

        int[] values = new int[]{
                porcentajes[0],
                porcentajes[0] + porcentajes[1],
                porcentajes[0] + porcentajes[1] + porcentajes[2],
                porcentajes[0] + porcentajes[1] + porcentajes[2] + porcentajes[3],
                porcentajes[0] + porcentajes[1] + porcentajes[2] + porcentajes[3] + porcentajes[4],
                porcentajes[0] + porcentajes[1] + porcentajes[2] + porcentajes[3] + porcentajes[4] + porcentajes[5],
                porcentajes[0] + porcentajes[1] + porcentajes[2] + porcentajes[3] + porcentajes[4] + porcentajes[5] + porcentajes[6]
        };

        // val banda = Array<Musician>(50){Cantante("", 2, TonoType.BAJO)}
        Musician[] banda = new Musician[50];

        for (int i = 0; i < banda.length; i++){
            String nombre = getNombreRandom();
            int experiencia = (int)(Math.random() * 40);
            int rdn = (int) (Math.random() * 100);
            if (rdn > 0 && rdn < values[0])
            {
                banda[i] = new Cantante(nombre, experiencia, getTonoRandom());
            }else if (rdn > values[0] && rdn < values[1])
            {
                banda[i] = new Guitarrista(nombre, experiencia, getGuitarraRandom());
            }else if (rdn > values[1] && rdn < values[2])
            {
                banda[i] = new Bajista(nombre, experiencia, (int)(Math.random() * 47));
            }else if (rdn > values[2] && rdn < values[3])
            {
                banda[i] = new Teclista(nombre, experiencia, (int)(Math.random() * 10));
            }else if (rdn > values[3] && rdn < values[4])
            {
                banda[i] = new Percusionista(nombre, experiencia, getPercussionRandom());
            }else if (rdn > values[4] && rdn < values[5])
            {
                banda[i] = new Trompetista(nombre, experiencia, (int)(Math.random() * 25));
            }else if (rdn > values[5] && rdn < values[6])
            {
                banda[i] = new CantanteGuitarrista(nombre, experiencia, getTonoRandom(), getGuitarraRandom());
            }else {
                banda[i] = new Multinstrumentista(nombre, experiencia, (int)(Math.random() * 10), getPercussionRandom(), (int)(Math.random() * 10));
            }
        }
        return banda;
    }
}

File: banda/SergioDeIscar-banda-kotlin/SergioDeIscar-banda-kotlin/src/main/kotlin/factories/MusicianFactory.kt
package bandamusical

import bandamusical.models.*

class MusicianFactory {
    companion object {
        private var instance: MusicianFactory? = null
        fun getInstance(): MusicianFactory {
            if (instance == null) {
                instance = MusicianFactory()
            }
            return instance!!
        }
    }

    private fun getNombreRandom():String{
        return arrayOf("Pepe", "Juan", "Ana", "Pedro", "María").random()
    }

    fun getBandaRandom(porcentajes: IntArray = intArrayOf(20, 10, 10, 10, 15, 15, 15, 5)): Array<Musician>{
        require(porcentajes.size == 8){"Error: porcentajes ha de tener un tamaño de 8"}
        require(porcentajes.all { it > 0 }){"Error: ninguno valor de porcentajes puede sr menor o igual a 0"}
        require(porcentajes.sum() == 100){"Error: la suma de todos los valores de porcentajes a de ser igual a 100"}
        val values = arrayOf(
            porcentajes[0],
            porcentajes[0] + porcentajes[1],
            porcentajes[0] + porcentajes[1] + porcentajes[2],
            porcentajes[0] + porcentajes[1] + porcentajes[2] + porcentajes[3],
            porcentajes[0] + porcentajes[1] + porcentajes[2] + porcentajes[3] + porcentajes[4],
            porcentajes[0] + porcentajes[1] + porcentajes[2] + porcentajes[3] + porcentajes[4] + porcentajes[5],
            porcentajes[0] + porcentajes[1] + porcentajes[2] + porcentajes[3] + porcentajes[4] + porcentajes[5] + porcentajes[6]
        )

        // val banda = Array<Musician>(50){Cantante("", 2, TonoType.BAJO)}
        val banda: Array<Musician> by lazy {
            Array(50) { Cantante("", 2, TonoType.BAJO) }
        }

        for (i in banda.indices){
            val nombre = getNombreRandom()
            val experiencia = (0..40).random()
            banda[i] = when((0..100).random()){
                // Cantante
                in 0 until  values[0] -> Cantante(nombre, experiencia, TonoType.values().random())

                // Guitarrista
                in values[0] until values[1]-> Guitarrista(nombre, experiencia, GuitarraType.values().random())

                // Bajista
                in values[1] until values[2] -> Bajista(nombre, experiencia, (1..47).random())

                // Teclista
                in values[2] until values[3] -> Teclista(nombre, experiencia, (1..10).random())

                // Percusionista
                in values[3] until values[4] -> Percusionista(nombre, experiencia, PercussionType.values().random())

                // Trompetista
                in values[4] until values[5] -> Trompetista(nombre, experiencia)

                // CantanteGuitarrista
                in values[5] until values[6] -> CantanteGuitarrista(nombre, experiencia,
                    TonoType.values().random(), GuitarraType.values().random())

                // Multinstrimentalista
                else -> Multinstrumentista(nombre, experiencia, (1..47).random(),
                    PercussionType.values().random(), (1..5).random(),)
            }
        }
        return banda
    }
}